picasso/src/main/java/com/squareup/picasso3/RequestHandler.java

```diff
@@ -209,18 +221,49 @@ static void calculateInSampleSize(int reqWidth, int reqHeight, int width, int he
             : Math.min(heightRatio, widthRatio);
       }
     }
-    options.inSampleSize = sampleSize;
-    options.inJustDecodeBounds = false;
+    return sampleSize;
   }
 
   /**
    * Decode a byte stream into a Bitmap. This method will take into account additional information
    * about the supplied request in order to do the decoding efficiently (such as through leveraging
    * {@code inSampleSize}).
    */
-  static Bitmap decodeStream(Source source, Request request) throws IOException {
+  static Bitmap decodeStream(Source source, final Request request) throws IOException {
     BufferedSource bufferedSource = Okio.buffer(source);
 
+    if (Build.VERSION.SDK_INT >= 28) {
+      return decodeStreamP(request, bufferedSource);
+    }
+
+    return decodeStreamPreP(request, bufferedSource);
+  }
+
+  @TargetApi(28)
+  @SuppressLint("Override")
+  private static Bitmap decodeStreamP(final Request request, BufferedSource bufferedSource)
+      throws IOException {
+    ImageDecoder.Source imageSource =
+        ImageDecoder.createSource(ByteBuffer.wrap(bufferedSource.readByteArray()));
```

Comment:
Ugh

Reply:
I know...

Reply:
Looks like there's a perfectly valid, yet hidden overload here: https://android.googlesource.com/platform/frameworks/base/+/refs/heads/master/graphics/java/android/graphics/ImageDecoder.java#900

🤔

Reply:
why does ImageDecoder not have a streaming API?!

Reply:
It needs to read multiple times and have random access